### PR TITLE
Support scrooge-generator compiler flags in thrift_library rule

### PR DESCRIPTION
**Pull request by cattibrie**
### Description
Add _compiler_args_ argument to  **thrift_library** rule.
Use **ScroogeConfig, ScroogeOptionParser** from com.twitter.scrooge to parse scrooge compiler arguments along with other flags passed in ScroogeWorker.scala.
Update _Compiler class_ to be up-to-date with the Compiler class in com.twitter.scrooge.

### Motivation
It was impossible to pass compiler arguments and other flags to the scrooge compiler.

### CODE REVIEW

**Comment by liucijus** [comment]
@blorente,  @wisechengyi I would appreciate your help reviewing this one

File: twitter_scrooge/twitter_scrooge.bzl [hunk 1]
```diff
@@ -165,15 +171,16 @@ def _generate_jvm_code(ctx, label, compile_thrifts, include_thrifts, jar_output,
...
     worker_content = "{output}\n{paths}\n{flags}".format(
         output = jar_output.path,
         paths = path_content,
-        flags = worker_arg_pad + ":".join([
-            # always add finagle option which is a no-op if there are no services
-            # we could put "include_services" on thrift_info, if needed
-            "--with-finagle",
-            "--language={}".format(language),
-        ]),
+        # we could put "include_services" on thrift_info, if needed
```
Comment (blorente) [inline twitter_scrooge/twitter_scrooge.bzl:182]: I think the top half of this comment is in the compiler scala code, is this a leftover?

File: test/src/main/scala/scalarules/test/twitter_scrooge/string_in_filename_in_jar.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+if jar tf ${jar_file} | grep -q $3 ; then
+  if test "$should_be_in_file" = "true" ; then
+    exit 0
+  else
+    echo "ERROR: Found string $3 in ${jar_file}, when we were expecting not to find it."
+    exit 1
+  fi
+else
+  if test "$should_be_in_file" = "true" ; then
+    echo "ERROR: Not found string $3 in ${jar_file}, when we were expecting to find it."
```
Comment (blorente) [inline test/src/main/scala/scalarules/test/twitter_scrooge/string_in_filename_in_jar.sh:35]:
```suggestion
    echo "ERROR: Not found file $3 in ${jar_file}, when we were expecting to find it."
```

I think it makes it easier to undersand.

File: test/src/main/scala/scalarules/test/twitter_scrooge/string_in_filename_in_jar.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+fi
+
+dir="$(dirname $4)"
+jar_file="$dir/$2"
+
+if jar tf ${jar_file} | grep -q $3 ; then
+  if test "$should_be_in_file" = "true" ; then
+    exit 0
+  else
+    echo "ERROR: Found string $3 in ${jar_file}, when we were expecting not to find it."
```
Comment (blorente) [inline test/src/main/scala/scalarules/test/twitter_scrooge/string_in_filename_in_jar.sh:30]:
```suggestion
    echo "ERROR: Found file $3 in ${jar_file}, when we were expecting not to find it."
```

File: test/src/main/scala/scalarules/test/twitter_scrooge/string_in_filename_in_jar.sh [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+if test "$should_be_in_file" != "true" -a "$should_be_in_file" != "false" ; then
+  echo "ERROR: Please use only (\"true\" or \"false\") to specify whether you need the substring to be in the file."
+  echo "Refer to test/src/main/scala/scalarules/test/twitter_scrooge/string_in_jar_file.sh for documentation."
+  exit 1
+fi
+
+dir="$(dirname $4)"
+jar_file="$dir/$2"
+
+if jar tf ${jar_file} | grep -q $3 ; then
```
Comment (blorente) [inline test/src/main/scala/scalarules/test/twitter_scrooge/string_in_filename_in_jar.sh:26]: nit: If you extract this to a variable like `file_is_in_jar` it almost makes bash into a legible language.

File: test/src/main/scala/scalarules/test/twitter_scrooge/BUILD [hunk 5]
```diff
@@ -344,4 +351,28 @@ sh_test(
     data = [":justscrooges_java"],
 )
 
+sh_test(
```
Comment (blorente) [inline test/src/main/scala/scalarules/test/twitter_scrooge/BUILD:354]: Thank you for adding these tests!

**Review by blorente — APPROVED** [review]
LGTM 👍 Apart from a couple of nits, it's a great change to see upstreamed.
@liucijus For a bit more context, we've been using this internally for a while now, but we were blocked on the scrooge version upgrade to upstream it.

File: src/scala/io/bazel/rules_scala/scrooge_support/Compiler.scala [hunk 6]
```diff
@@ -33,8 +33,9 @@ package io.bazel.rules_scala.scrooge_support
 
 import com.twitter.scrooge._
 import com.twitter.scrooge.ast.Document
-import com.twitter.scrooge.backend.{GeneratorFactory, ScalaGenerator, ServiceOption}
+import com.twitter.scrooge.backend.{GeneratorFactory, ScalaGenerator}
 import com.twitter.scrooge.frontend.{FileParseException, TypeResolver, ThriftParser, Importer, MultiImporter, ZipImporter}
```
Comment (liucijus) [inline src/scala/io/bazel/rules_scala/scrooge_support/Compiler.scala:37]: nit: as you're touching this file, please do some cleanup too: seems that Importer, MultiImporter, ZipImporter are unused imports

File: src/scala/io/bazel/rules_scala/scrooge_support/Compiler.scala [hunk 7]
```diff
@@ -66,46 +64,34 @@ object CompilerDefaults {
...
-      if (verbose) {
+      if (config.verbose) {
         println("+ Writing file mapping to %s".format(path))
       }
       new FileWriter(file)
     }
 
     val allJars: List[File] =
-      ((includeJars.toList) ::: (compileJars.toList))
+      ((includeJars) ::: (compileJars))
```
Comment (liucijus) [inline src/scala/io/bazel/rules_scala/scrooge_support/Compiler.scala:91]: nit: please remove redundant braces around vals

**Review by liucijus — CHANGES_REQUESTED** [review]
LGTM, could you address my nit comments?

File: src/scala/scripts/ScroogeWorker.scala [hunk 8]
```diff
@@ -86,10 +73,19 @@ object ScroogeWorker extends Worker.Interface {
       sys.error("onlyTransitiveThriftSrcs and immediateThriftSrcs should " +
         s"have not intersection, found: ${intersect.mkString(",")}")
 
+    // To preserve current default behaviour.
+    val defaultConfig = ScroogeConfig(
+      destFolder = scroogeOutput.toString,
+      includePaths = onlyTransitiveThriftSrcJars ++ remoteSrcJars,
+      // always add finagle option which is a no-op if there are no services
```
Comment (cattibrie) [inline src/scala/scripts/ScroogeWorker.scala:80]: I wonder if this default is needed when this flag can be passed in `compiler_args` field in **thrift_library** rule?

Comment (blorente) [reply src/scala/scripts/ScroogeWorker.scala:80]: People could pass it from outside, but flipping the default value from true to false would be a semantic (and probably breaking) change.
I think that's a great avenue to explore if you're interested, but I'd probably do it as a follow-up, and not part of this big diff.

**Review by liucijus — APPROVED** [review]

**Review by ittaiz — APPROVED** [review]

**Comment by liucijus** [comment]
Thanks, @cattibrie!

**Pull request merged**